package-name-map/src/test/package-name-map_test.ts

```diff
@@ -0,0 +1,226 @@
+        assert.throws(() => map.resolve('app/foo.js', referrerURL));
+      });
+    });
+
+    suite('path_prefix', () => {
+      const map = new PackageNameMap({
+        path_prefix: '/node_modules',
+        packages: {
+          moment: {
+            main: 'moment.js',
+          },
+          lodash: {
+            path: 'lodash-es',
+            main: 'lodash.js',
+          },
+          '@polymer/polymer': {
+            path: '@polymer/polymer',
+            main: 'polymer.js',
+          },
+        },
+      });
+
+      test('resolves package name for a package with only a main', () => {
+        assert.equal(
+          map.resolve('moment', referrerURL),
+          '/node_modules/moment/moment.js'
+        );
+      });
+
+      test('resolves package name for a package with a path and main', () => {
+        assert.equal(
+          map.resolve('lodash', referrerURL),
+          '/node_modules/lodash-es/lodash.js'
+        );
+      });
+
+      test('resolves a submodule for package with a path and main', () => {
+        assert.equal(
+          map.resolve('lodash/bar.js', referrerURL),
+          '/node_modules/lodash-es/bar.js'
+        );
+      });
+
+      test('resolves a "scoped" package name', () => {
+        assert.equal(
+          map.resolve('@polymer/polymer', referrerURL),
+          '/node_modules/@polymer/polymer/polymer.js'
+        );
+      });
+    });
+
+    suite('scopes', () => {
+      const map = new PackageNameMap({
+        packages: {
+          moment: {
+            main: 'fail.js',
+          },
+          lodash: {
+            path: '/node_modules/lodash-es',
+            main: 'lodash.js',
+          },
+        },
+        scopes: {
+          '/node_modules/lodash-es': {
+            path_prefix: 'node_modules',
+            packages: {
+              moment: {
+                main: 'moment.js',
+              },
+            },
+            scopes: {
+              subpackage: {
+                packages: {
+                  'not-moment': {
+                    main: 'index.js',
+                  },
+                },
+              },
+            },
+          },
+        },
+      });
+
+      test('resolves package names in scopes', () => {
+        assert.equal(
+          map.resolve('moment', '/node_modules/lodash-es/lodash.js'),
+          '/node_modules/lodash-es/node_modules/moment/moment.js'
+        );
+      });
+
+      test('resolves package names in outer scopes', () => {
+        assert.equal(
+          map.resolve('lodash', '/node_modules/lodash-es/lodash.js'),
+          '/node_modules/lodash-es/lodash.js'
+        );
+      });
+
+      test('resolves package names in outer scopes from deeper scopes', () => {
+        assert.equal(
+          map.resolve('moment', '/node_modules/lodash-es/subpackage'),
+          '/node_modules/lodash-es/node_modules/moment/moment.js'
+        );
+      });
+    });
+  });
+});
+
+suite('isPathSegmentPrefix', () => {
```

Comment:
Maybe best in a separate file since it's not something we'd eventually port to web platform tests?